src/plugins/liveobjects/livecounter.ts

```diff
@@ -1,15 +1,100 @@
 import { LiveObject, LiveObjectData } from './liveobject';
+import { LiveObjects } from './liveobjects';
+import { StateCounter, StateCounterOp, StateOperation, StateOperationAction } from './statemessage';
 
 export interface LiveCounterData extends LiveObjectData {
   data: number;
 }
 
 export class LiveCounter extends LiveObject<LiveCounterData> {
+  constructor(
+    liveObjects: LiveObjects,
+    private _created: boolean,
+    initialData?: LiveCounterData | null,
+    objectId?: string,
+  ) {
+    super(liveObjects, initialData, objectId);
+  }
+
   value(): number {
     return this._dataRef.data;
   }
 
+  /**
+   * @internal
+   */
+  isCreated(): boolean {
+    return this._created;
+  }
+
+  /**
+   * @internal
+   */
+  setCreated(created: boolean): void {
+    this._created = created;
+  }
+
+  /**
+   * @internal
+   */
+  applyOperation(op: StateOperation): void {
+    if (op.objectId !== this.getObjectId()) {
+      throw new this._client.ErrorInfo(
+        `Cannot apply state operation with objectId=${op.objectId}, to this LiveCounter with objectId=${this.getObjectId()}`,
+        50000,
+        500,
+      );
+    }
+
+    switch (op.action) {
+      case StateOperationAction.COUNTER_CREATE:
+        if (this.isCreated()) {
+          // skip COUNTER_CREATE op if this counter is already created
+          break;
+        }
+
+        if (this._client.Utils.isNil(op.counter)) {
+          this._throwNoPayloadError(op);
```

Comment:
hm, a bit unexpected as conceptually the same properties from the [state _object_ message](https://github.com/ably/realtime/blob/main/protocol/state.proto#L77-L79) will always  be present depending if it's a counter/map object. But I guess in that case the presence of an object is a discrimination factor in itself, so you must have one of them to distinguish the indended operation, even if it is empty.

Reply:
anyway, based on your comment it seems that it is logically correct to process a COUNTER_CREATE/MAP_CREATE op without a corresponding object and just create an empty one.
will do that, and we can figure out if we want to enforce including this field when empty separately.

Reply:
> anyway, based on your comment it seems that it is logically correct to process a COUNTER_CREATE/MAP_CREATE op without a corresponding object and just create an empty one

correct, see: https://github.com/ably/realtime/blob/main/go/realtime/lib/state/map.go#L105

> hm, a bit unexpected as conceptually the same properties from the [state object message](https://github.com/ably/realtime/blob/main/protocol/state.proto#L77-L79) will always be present depending if it's a counter/map object. But I guess in that case the presence of an object is a discrimination factor in itself, so you must have one of them to distinguish the indended operation, even if it is empty.

If we have a `StateOperation` message, `*_CREATE` ops will use the `map` or `counter` fields to specify the initial value (which is optional). Other operations (i.e. `MAP_SET/REMOVE`, `COUNTER_INC`) will use the `mapOp` and `counterOp` fields. It is never the case that both `map` and `mapOp` are set together - we know entirely from the `action` which fields we expect to be set.

Reply:
also will update here https://github.com/ably/ably-js/pull/1897/files#diff-08699f48600bb9488e2ac20a925d42439ab96df2cc16120e33cdcb4e1d4886fcR94-R106 -- it covers creating objects if they didn't exist before in the pool

Reply:
> If we have a StateOperation message, *_CREATE ops will use the map or counter fields to specify the initial value (which is optional). Other operations (i.e. MAP_SET/REMOVE, COUNTER_INC) will use the mapOp and counterOp fields. It is never the case that both map and mapOp are set together - we know entirely from the action which fields we expect to be set.

understandable, I was referring to the [`StateObject` type](https://github.com/ably/realtime/blob/main/protocol/state.proto#L71) which has conceptually the same `map` and `counter` fields, but one of them is always present there, since there would be no way to know which type of the object to create otherwise

Reply:
made suggested changes, see compare https://github.com/ably/ably-js/compare/0a70bb1faa7145bcef36018758bd921069609c2d..a9f916939a12752f478c21972a08dd732bf729ac